iotdb-core/datanode/src/main/java/org/apache/iotdb/db/utils/datastructure/MergeSortAlignedTVListIterator.java

```diff
@@ -0,0 +1,216 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.utils.datastructure;
+
+import org.apache.tsfile.enums.TSDataType;
+import org.apache.tsfile.file.metadata.enums.TSEncoding;
+import org.apache.tsfile.read.TimeValuePair;
+import org.apache.tsfile.read.reader.IPointReader;
+import org.apache.tsfile.utils.BitMap;
+import org.apache.tsfile.utils.TsPrimitiveType;
+
+import java.io.IOException;
+import java.util.List;
+
+public class MergeSortAlignedTVListIterator implements IPointReader {
+  private final AlignedTVList.AlignedTVListIterator[] alignedTvListIterators;
+  private final int columnNum;
+
+  private boolean probeNext = false;
+  private boolean hasNext = false;
+
+  private final int[] alignedTvListOffsets;
+
+  // Remember the selected columns for prepareNext
+  // column index -> [selectedTVList, selectedIndex]
+  private final int[][] columnAccessInfo;
+
+  private long time;
+  private final BitMap bitMap;
+
+  public MergeSortAlignedTVListIterator(
+      List<AlignedTVList> alignedTvLists,
+      List<TSDataType> tsDataTypes,
+      List<Integer> columnIndexList,
+      Integer floatPrecision,
+      List<TSEncoding> encodingList,
+      boolean ignoreAllNullRows) {
+    this.alignedTvListIterators = new AlignedTVList.AlignedTVListIterator[alignedTvLists.size()];
+    for (int i = 0; i < alignedTvLists.size(); i++) {
+      alignedTvListIterators[i] =
+          alignedTvLists
+              .get(i)
+              .iterator(
+                  tsDataTypes, columnIndexList, ignoreAllNullRows, floatPrecision, encodingList);
+    }
+    this.alignedTvListOffsets = new int[alignedTvLists.size()];
+    this.columnNum = tsDataTypes.size();
+    this.columnAccessInfo = new int[columnNum][];
+    for (int i = 0; i < columnAccessInfo.length; i++) {
+      columnAccessInfo[i] = new int[2];
+    }
+    this.bitMap = new BitMap(columnNum);
+  }
+
+  public MergeSortAlignedTVListIterator(
+      AlignedTVList.AlignedTVListIterator[] alignedTvListIterators, int columnNum) {
+    this.alignedTvListIterators =
+        new AlignedTVList.AlignedTVListIterator[alignedTvListIterators.length];
+    for (int i = 0; i < alignedTvListIterators.length; i++) {
+      this.alignedTvListIterators[i] = alignedTvListIterators[i].clone();
+    }
+    this.alignedTvListOffsets = new int[alignedTvListIterators.length];
+    this.columnNum = columnNum;
+    this.columnAccessInfo = new int[columnNum][];
+    for (int i = 0; i < columnAccessInfo.length; i++) {
+      columnAccessInfo[i] = new int[2];
+    }
+    this.bitMap = new BitMap(columnNum);
+  }
+
+  private void prepareNextRow() {
+    time = Long.MAX_VALUE;
+    for (int i = 0; i < alignedTvListIterators.length; i++) {
```

Comment:
Add a comment:
Here it is a linear scan for all tvlists. This is because I want keep it simple when tvlist_sort_threshold is 0 (no tvlist handover). I assume it should not set tvlist_sort_threshold a very small value, so there are no large number of tvlist in memchunk. If we should consider the worst case, I can add the heap implementation for merge sort.

Reply:
Done